docs/gob.1
.TH GOB  "1"
.SH NAME
gob \- block-based backup utility
.SH SYNOPSIS
.B gob [\-\-version] <COMMAND> [ARGS...]
.SH DESCRIPTION
gob is a block-based backup utility.
It tries to be as simple as possible while doing everything necessary to create deduplicated backups.
To do so, it strictly follows the Unix principle of doing one thing per command.
.sp
The main idea of gob is based on the notion of block storages and indices.
The block storage is using a content-adressable file system inspired by git.
gob will read blocks up to a specified maximum block size, hash this block and then put it into the block storage using the hash as name.
So when reading the same block twice, it will only get stored once.
.sp
The block storage is only the first part, as the restoration process needs to be able to read blocks in the correct order, again.
To do so, the chunking program will output a list of hashes of the blocks read, which is called an "index".
Using this list, it becomes trivial to restore the complete data by simply concatenating all blocks in the order given by the index.
.SH OPTIONS
\-\-version
.RS 4
Show version and build information.
.RE
.SH COMMANDS
gob-cat(1)
.RS 4
Read from a block store.
.RE
.PP
gob-chunk(1)
.RS 4
Store data in a block store.
.RE
.PP
gob-fsck(1)
.RS 4
Check consistency of a block store.
.RE


CHANGELOG
gob v0.8
========

Changes
-------

- Standard output is now being checked for errors when exiting to
  verify that e.g. the index has been completely written.

- Continuous integration has been updated from Trusty to Xenial.
  Another build job for Bionic has been added.

- A lot of warnings have been fixed in order to enable the
  continuous integration to build with `-Werror`.

- Tests have been refactored to avoid any use of pipes which may
  swallow error codes.

- An error has been fixed that caused `warn()` to call exit(3P),
  causing gob-fsck(1) to exit on the first warning.

- Added an editorconfig file.

gob v0.7
========

- The build system has been migrated from Makefiles to Meson.

- Top-level source files and manpages have been moved into a new
  "src/" and "docs/" subdirectory, respectively.

- Support for encryption has been completely removed. Its
  security cannot be guaranteed and better solutions exist by
  e.g. writing the backup store into a LUKS volume.

- The hex conversion and blake2b hash functions have been
  included into our sources. This allowed removing the dependency
  on libsodium altogether.

gob v0.6
========

Changes
-------

- Build instructions and tests have been adjusted to not include
  non-POSIX instructions. This enables support for OpenBSD.

- When writing a block into the block store, gob will now first
  write everything into a temporary file and then rename that
  file to have the final name. This will ensure that gob will
  never put partially written blocks at a final block path.

- Travis CI has been fixed to check the signature of packages
  installed via Bintray.

gob v0.5
========

Additions
---------

- A new tool "gob-fsck" has been added that checks an object
  storage for consistency.

- A CI job has been added to do static analysis on gob's sources
  via Coverity.

Changes
-------


- Previously, gob-chunk would just error out and leave the
  partially written block in the object storage. Now, it instead
  tries to unlink it first and only then abort.

- The block storage has been abstracted into its own structure
  and set of functions.

gob v0.4
========

Changes
-------

- Rename from "sb" to "gob". All binaries now have a "gob-"
  prefix instead of a "sb-" prefix.

- The block storage now has a version associated with it. This
  makes sure that gob will not try to write to block stores which
  have a different version, especially in the case of
  backwards-incompatible changes.

sb v0.3
=======

Additions
---------

- Travis has been set up to build and test changes.

Changes
-------

- Data that is output by sb-encrypt is using fixed-length blocks
  now, with the actual data's size being encoded as part of the
  ciphertext.

- Switch to a secure scheme for generating encryption nonces.
  Previously, nonces were simply the index of the current block
  and as such completely deterministic. Nonces are now derived
  from the hash of the index concatenated with the plain-text
  data. This generates consistent nonces for the same block while
  still being unpredictable from the outside without knowing the
  plain text of the block.

sb v0.2
=======

Changes
-------

- Blocks are now allocated dynamically on the heap. Previously,
  all blocks were allocated on the stack, which could have hit
  the maximum stack size in case where the user has specified a
  large block size via config.h.

- The default block size was increased to 4MB. The previous
  default of 4KB resulted in huge block lists and tons of block
  files.

sb v0.1
=======

This is the initial release of the simple backup tool. It
provides the following tools:

- sb-chunk writes data from standard input into a block store.
  Chunks are being deduplicated based on the hash of the data
  that is being written into the block store. The order of blocks
  written into the blockstore are written to standard output.

- sb-cat will read a list of blobs and successively read each of
  these blocks from a given block store to standard output.

- sb-encrypt will read data from standard input and write it
  to standard output, encrypting it with a key.

- sb-decrypt will read data from standard input and write it to
  standard output, decrypting it with a key.


src/gob.c
/*
 * Copyright (C) 2019 Patrick Steinhardt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "common.h"

#define ARRAY_SIZE(x) (sizeof(x)/sizeof(*x))

static struct {
    int (*fn)(int argc, const char *argv[]);
    const char *name;
    const char *description;
} commands[] = {
    { gob_cat,   "cat",   "Concatenate chunks" },
    { gob_chunk, "chunk", "Chunk and store data" },
    { gob_fsck,  "fsck",  "Check consistency of a store"  },
    { gob_init,  "init",  "Initialize a new store"  },
};

int main(int argc, const char *argv[])
{
    size_t i;

    if (argc <= 1) {
        fprintf(stderr, "USAGE: %s [--version] <COMMAND>\n\n", argv[0]);
        fprintf(stderr, "The following commands are available:\n\n");

        for (i = 0; i < ARRAY_SIZE(commands); i++) {
            fprintf(stderr, "\t%s\t%s\n", commands[i].name, commands[i].description);
        }

        return 1;
    } else if (!strcmp(argv[1], "--version")) {
        printf("%s version "GOB_VERSION"\n\n"
               "block size: %d\n"
               "hash size:  %d\n", argv[0], BLOCK_LEN, HASH_LEN);
        return 0;
    }

    for (i = 0; argc > 1 && i < ARRAY_SIZE(commands); i++) {
        if (strcmp(argv[1], commands[i].name))
            continue;
        memmove(argv + 1, argv + 2, sizeof(char *) * (unsigned) argc - 2);
        return commands[i].fn(argc - 1, argv);
    }

    if (i == ARRAY_SIZE(commands))
        fprintf(stderr, "%s: '%s' is not a known gob command\n", argv[0], argv[1]);

    return 1;
}
